software/statemachine.py

```diff
@@ -0,0 +1,233 @@
+        if normalizedYDistanceFromBottom > 0.4:
+            self.currentState = State.FIND_BALL
+            return
+
+        normalizedXDistanceFromCenter = (ballXCoord - (self.imageCenterPoint)) / self.imageWidth
+        basketMaxDistanceFromCenter = 15 #in pixels
+        #additional offset
+        basketOffset = -5
+
+        #check if go into throwing state
+        if self.throwIntoBlue:
+            if abs(self.imageData.basket_b.x - self.imageCenterPoint + basketOffset) < basketMaxDistanceFromCenter:
+                #if orbit is inaccurate, correct position once more
+                if abs(normalizedXDistanceFromCenter) > 0.15 or normalizedYDistanceFromBottom > 0.3:
+                    self.currentState = State.GO_TO_BALL
+                    self.orbitDone = True
+                    return
+                else:
+                    self.currentState = State.THROW
+                    return
+        else:
+            if abs(self.imageData.basket_m.x - self.imageCenterPoint + basketOffset) < basketMaxDistanceFromCenter:
+                if abs(normalizedXDistanceFromCenter) > 0.15 or normalizedYDistanceFromBottom > 0.3:
+                    self.currentState = State.GO_TO_BALL
+                    self.orbitDone = True
+                    return
+                else:
+                    self.currentState = State.THROW
+                    return
+
+
+        #speed calculations
+        trajectorySpeed = 0.3        
+        rotBase = 0.3
+        rotSpeedMultiplier = -4
+        rotSpeed = rotBase + normalizedXDistanceFromCenter * rotSpeedMultiplier
+
+        #changing orbit speed 
+        slowOrbitZone = 100
+        seeBasketSpeed = 0.2
+        closeBasketSpeed = 0.05
+        if self.throwIntoBlue:
+            #if i see basket, move slower
+            if self.imageData.basket_b.exists:
+                trajectorySpeed = seeBasketSpeed
+                #if basket is close to center, move even slower
+                if abs(self.imageData.basket_b.x - self.imageCenterPoint) < slowOrbitZone:
+                    trajectorySpeed = closeBasketSpeed
+                #if basket is on right side of screen, move clockwise
+                if self.imageData.basket_b.x > (self.imageCenterPoint):
+                    trajectorySpeed = -trajectorySpeed
+                    rotSpeed = -rotSpeed
+        else:
+            if self.imageData.basket_m.exists:
+                trajectorySpeed = seeBasketSpeed
+                if abs(self.imageData.basket_m.x - self.imageCenterPoint) < slowOrbitZone:
+                    trajectorySpeed = closeBasketSpeed
+                if self.imageData.basket_m.x > (self.imageCenterPoint):
+                    trajectorySpeed = -trajectorySpeed
+                    rotSpeed = -rotSpeed
+
+        self.robot.move(trajectorySpeed, 0, rotSpeed)
+
+    #averages pixel distances of a 5x5 square from depth_image
+    def basketDist(self, x, y):
+        distanceSum = 0
+        for i in range(-2, 3):
```

Comment:
You can select a range of pixels using numpy and use the np.average method for that

Reply:
Good to know